components/kubernetes/component.go

```diff
@@ -0,0 +1,99 @@
+package kubernetes
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/DataDog/test-infra-definitions/common/config"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+type Importable interface {
+	Key() string
+	Import(output auto.OutputMap) error
+}
+
+// The generic component resource that be created by any provider (kind, EKS, etc.)
+// We could apply the same thing for VM with AWS VMs, Azure VMs, and so on
+type KubernetesCluster struct {
+	pulumi.ResourceState
+	exportName string
+
+	KubeConfig pulumi.StringOutput `export:"kubeConfig"`
+}
+
+// The target type that's going to be deserialized on the other side.
+// We could unit test the matching if required.
+// The E stands for exported, probably can find something better
+type EKubernetesCluster struct {
+	Key string
+
+	KubeConfig string `json:"kubeConfig"`
+}
+
+// func NewEKubernetesCluster could be implemented with clusterName parameter to be able
+// to deserialize when Export is done inside predefined runFunc.
+// We could also implement a scan of OutputMap and the relevant key auto if only one if this kind present.
+
+// We can have a base class that implements Import and Key as there's no value in copy pasting
+func (ek *EKubernetesCluster) Import(output auto.OutputMap) error {
+	if ek.Key == "" {
+		return errors.New("missing key")
+	}
+
+	val, found := output[ek.Key]
+	if !found {
+		return errors.New("output not found")
+	}
+
+	// it *might* happen that val.Value is a map[interface{}]interface{} or something, which
+	// would mean that we need to implement our own reflection as well to import.
+	return json.Unmarshal(val.Value.([]byte), ek)
+}
+
+// Boilerplate component resource creation
+func newClusterComponent(e *config.CommonEnvironment, name string, opts ...pulumi.ResourceOption) (*KubernetesCluster, error) {
+	comp := &KubernetesCluster{
+		// as we components we can guarantee key unicity without asking user to provide name again
+		exportName: e.CommonNamer.ResourceName("dd", "kubeCluster", name),
+	}
+
+	if err := e.Ctx.RegisterComponentResource("dd:kubeCluster", e.CommonNamer.ResourceName("kind", name), comp, opts...); err != nil {
+		return nil, err
+	}
+
+	return comp, nil
+}
+
+func (k *KubernetesCluster) Export(e *config.CommonEnvironment, exportedObj *EKubernetesCluster) {
+	// To reproduce the current cross-program assignment in `datadog-agent`, not technically required.
+	if exportedObj != nil {
+		exportedObj.Key = k.exportName
+	}
+
+	// This can be easily replaced by reflection traversal but not enough time rn
+	e.Ctx.Export(k.exportName, pulumi.JSONMarshal(pulumi.ToMapOutput((map[string]pulumi.Output{
+		"kubeConfig": k.KubeConfig,
+	}))))
+}
+
+// // Usage in datadog-agent
```

Comment:
Can you elaborate a little the usage in datadog-agent for example what will you keep, what will you remove?

Reply:
If I understand it well, a client in `datadog-agent` would get a specific `test-infra-definitions` deserialiser, and would call the `deserialise` at init. What could be confusing is how to keep `Export`, but is defined thanks to the environment that exists only in a pulumi context. Sane fir `Import`, it can be called only with an UpResult, so that would be called by the stack definition. 

Each client would require to have a `Deserialiser` component, and then the environment initialiser would call each component `Import`.